src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.BatchingDocumentCollection.cs

```diff
@@ -439,55 +452,66 @@ await _project._projectSystemProjectFactory.ApplyBatchChangeToWorkspaceAsync((so
-                {
-                    return;
-                }
+                return new AsyncBatchingWorkQueue<(string, string)>(
+                    TimeSpan.FromMilliseconds(200), // 200 chosen with absolutely no evidence whatsoever
+                    ProcessDynamicFileChangesAsync,
+                    EqualityComparer<(string, string)>.Default, // uses ordinal string comparison which is what we want
+                    _project._projectSystemProjectFactory.WorkspaceListener,
+                    _project._asynchronousFileChangeProcessingCancellationTokenSource.Token);
+            });
+
+            _dynamicFilesToRefresh.AddWork((projectSystemFilePath, workspaceFilePath));
+        }
+
+        private async ValueTask ProcessDynamicFileChangesAsync(ImmutableSegmentedList<(string projectSystemFilePath, string workspaceFilePath)> batch, CancellationToken cancellationToken)
+        {
+            foreach (var (projectSystemPath, workspaceFilePath) in batch)
+            {
+                DocumentId? documentId;
+                IDynamicFileInfoProvider? fileInfoProvider;
 
-                if (_documentPathsToDocumentIds.TryGetValue(workspaceFilePath, out var documentId))
+                using (await _project._gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
                 {
-                    // We create file watching prior to pushing the file to the workspace in batching, so it's
-                    // possible we might see a file change notification early. In this case, toss it out. Since
-                    // all adds/removals of documents for this project happen under our lock, it's safe to do this
-                    // check without taking the main workspace lock. We don't have to check for documents removed in
-                    // the batch, since those have already been removed out of _documentPathsToDocumentIds.
-                    if (_documentsAddedInBatch.Any(d => d.Id == documentId))
-                    {
+                    // If our project has already been removed, this is a stale notification, and we can disregard.
+                    if (_project.HasBeenRemoved)
                         return;
-                    }
 
-                    Contract.ThrowIfFalse(_documentIdToDynamicFileInfoProvider.TryGetValue(documentId, out var fileInfoProvider));
+                    if (!_documentPathsToDocumentIds.TryGetValue(workspaceFilePath, out documentId))
+                        return;
 
-                    _project._projectSystemProjectFactory.ApplyChangeToWorkspace(w =>
+                    if (!_documentIdToDynamicFileInfoProvider.TryGetValue(documentId, out fileInfoProvider))
+                        return;
+                }
+
+                // Now that we've got all our basic data, let's fetch the new document outside the lock, since this could be expensive.
+                var fileInfo = await fileInfoProvider.GetDynamicFileInfoAsync(
```

Comment:
Should we consider parallelizing the fetch of the dynamic files in the batch?  And then maybe we could also apply all the documents in the batch at once, avoiding needing to take the workspace lock for each one?